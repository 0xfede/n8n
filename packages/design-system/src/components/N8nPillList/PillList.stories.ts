import type { StoryFn } from '@storybook/vue3';
import { ref, watch } from 'vue';

import N8nPillList from './PillList.vue';

export default {
	title: 'Modules/PillList',
	component: N8nPillList,
	argTypes: {},
	parameters: {
		backgrounds: { default: '--color-background-light' },
	},
};

const Template: StoryFn = (args, { argTypes }) => ({
	setup: () => ({ args }),
	props: Object.keys(argTypes),
	// Generics make this difficult to type
	components: N8nPillList as never,
	template: '<n8n-pill-list v-bind="args"  />',
});

export const PillList = Template.bind({});
PillList.args = {
	modelValue: {
		propC: 'propC value that will be changed in the slot',
	},
	inputs: [
		{
			name: 'propA',
			tooltip: 'propA tooltip',
			default: 'propA default',
		},
		{
			name: 'propB',
			tooltip: 'propB tooltip',
			default: 'propB default',
		},
		{
			name: 'propC',
			tooltip: 'propC tooltip',
			default: 'propC default',
		},
	],
};
