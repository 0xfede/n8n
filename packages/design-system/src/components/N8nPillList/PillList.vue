<script
	setup
	lang="ts"
	generic="Value, Item extends { name: string; defaultValue: Value; tooltip?: string }"
>
import { computed } from 'vue';

defineSlots<{
	displayItem: (props: Item) => unknown;
}>();

type PillListProps = {
	inputs: Item[];
};

const props = withDefaults(defineProps<PillListProps>(), {
	inputs: () => [],
});

// Record<inputs[k].name, currentValue>
const selectedItems = defineModel<Record<string, Value>>({ required: true });
const inputMap = computed(() => Object.fromEntries(props.inputs.map((x) => [x.name, x] as const)));

const visiblePills = computed(() => {
	return props.inputs.filter((pill) => !selectedItems.value.hasOwnProperty(pill.name));
});

const sortedSelectedItems = computed(() => {
	return [...Object.keys(selectedItems.value).map((name) => inputMap.value[name])].sort((a, b) =>
		a.name[0] < b.name[0] ? -1 : 1,
	);
});

function addToSelectedItems(name: string) {
	selectedItems.value[name] = inputMap.value[name].defaultValue;
}

function removeFromSelectedItems(name: string) {
	delete selectedItems.value[name];
}
</script>

<template>
	<div>
		<div :class="$style.pillContainer">
			<span
				v-for="item in [...visiblePills].sort((a, b) => (a.name < b.name ? -1 : 1))"
				:key="item.name"
				:class="$style.pillCell"
				:data-test-id="`pill-list-pill-${item.name}`"
				@click="addToSelectedItems(item.name)"
			>
				+ Add a {{ item.name }}
			</span>
		</div>
		<div
			v-for="item in sortedSelectedItems"
			:key="item.name"
			:class="$style.slotComboContainer"
			:data-test-id="`pill-list-slot-${item.name}`"
		>
			<N8nIconButton
				type="tertiary"
				text
				:class="$style.slotRemoveIcon"
				size="mini"
				icon="trash"
				:data-test-id="`pill-list-remove-slot-${item.name}`"
				@click="removeFromSelectedItems(item.name)"
			/>
			<div :class="$style.slotContainer">
				<slot name="displayItem" v-bind="item"
					>Empty slot with v-bind: {{ JSON.stringify(item) }}</slot
				>
			</div>
		</div>
	</div>
</template>

<style lang="scss" module>
.slotComboContainer {
	display: flex;
	flex-direction: row;
	flex-grow: 1;
	margin-top: var(--spacing-2xs);
}

.slotContainer {
	flex-grow: 1;
}

.pillContainer {
	width: 100%;
	flex-wrap: wrap;
	display: flex;
}
.pillCell {
	display: flex;
	margin: var(--spacing-4xs) var(--spacing-3xs) 0 0;

	min-width: max-content;
	border-radius: var(--border-radius-base);
	font-size: small;
	background-color: var(--color-ndv-background);
	color: var(--text-color-dark);

	cursor: pointer;

	:hover {
		color: var(--color-primary);
	}
}
</style>
