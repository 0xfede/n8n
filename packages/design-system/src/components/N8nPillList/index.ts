import N8nPillList from './PillList.vue';

export default N8nPillList;
