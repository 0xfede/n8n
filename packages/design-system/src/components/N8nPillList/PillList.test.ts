import { fireEvent, render, waitFor } from '@testing-library/vue';

import N8nPillList from './PillList.vue';

const itemSize = 100;
const itemKey = 'id';
const items = [...(new Array(100) as number[])].map((_, index) => ({
	id: String(index),
	name: `Item ${index}`,
}));

describe('components', () => {
	describe('N8nRecycleScroller', () => {
		it('renders when empty', () => {
			const wrapper = render(N8nPillList, {
				props: {
					modelValue: {},
					inputs: [],
				},
			});
			expect(wrapper.html()).toMatchSnapshot();
		});

		it('renders one clickable element that can be added and removed', async () => {
			const wrapper = render(N8nPillList, {
				props: {
					modelValue: {},
					inputs: [{ name: 'propA', defaultValue: '' }],
				},
			});

			expect(wrapper.getByTestId('pill-list-pill-propA')).toBeInTheDocument();

			await fireEvent.click(wrapper.getByTestId('pill-list-pill-propA'));

			expect(wrapper.queryByTestId('pill-list-pill-propA')).not.toBeInTheDocument();
			expect(wrapper.getByTestId('pill-list-slot-propA')).toBeInTheDocument();

			await fireEvent.click(wrapper.getByTestId('pill-list-remove-slot-propA'));

			expect(wrapper.queryByTestId('pill-list-slot-propA')).not.toBeInTheDocument();
		});

		it('renders multiple elements with some pre-selected', () => {
			const wrapper = render(N8nPillList, {
				props: {
					modelValue: {
						propA: 'propA value',
					},
					inputs: [
						{ name: 'propA', defaultValue: '' },
						{ name: 'propB', defaultValue: '' },
						{ name: 'propC', defaultValue: '' },
					],
				},
			});

			expect(wrapper.queryByTestId('pill-list-pill-propA')).not.toBeInTheDocument();
			expect(wrapper.getByTestId('pill-list-pill-propB')).toBeInTheDocument();
			expect(wrapper.getByTestId('pill-list-pill-propC')).toBeInTheDocument();
			expect(wrapper.getByTestId('pill-list-slot-propA')).toBeInTheDocument();

			expect(wrapper.html()).toMatchSnapshot();
		});
	});
});
