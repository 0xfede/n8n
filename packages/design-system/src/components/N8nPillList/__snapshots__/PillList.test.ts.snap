// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components > N8nRecycleScroller > renders multiple elements with some pre-selected 1`] = `
"<div>
  <div class="pillContainer"><span class="pillCell" data-test-id="pill-list-pill-propB"> + Add a propB</span><span class="pillCell" data-test-id="pill-list-pill-propC"> + Add a propC</span></div>
  <div class="slotComboContainer" data-test-id="pill-list-slot-propA"><button class="button button tertiary mini text withIcon square slotRemoveIcon slotRemoveIcon" aria-live="polite" data-test-id="pill-list-remove-slot-propA"><span class="icon"><span class="n8n-text compact size-xsmall regular n8n-icon n8n-icon"><!----></span></span>
      <!--v-if-->
    </button>
    <div class="slotContainer">Empty slot with v-bind: {"name":"propA","defaultValue":""}</div>
  </div>
</div>"
`;

exports[`components > N8nRecycleScroller > renders when empty 1`] = `
"<div>
  <div class="pillContainer"></div>
</div>"
`;
